enhancements/ingress/route-secret-injection-for-external-certificate-management.md

```diff
@@ -0,0 +1,375 @@
+that have the valid annotation for `tls-secret-reference`. This controller will also update
+`.status` of the route to contain the latest information on the certificate in use.
+
+In scenarios where the user deletes the annotation, it is the controller's 
+responsibility to reset `.spec.tls` and this will be driven based on `.status` i.e. 
+if the status denotes `RouteCertificateManagedType` and since the required annotation is 
+not present this results in the `.spec.tls` being cleared out and the `.status` updated 
+to `RouteCertificateDefaultType`.
+
+If the user deletes the secret associated with the Route through the `route.openshift.io/tls-secret-reference`
+annotation, the secret-injector controller will delete this annotation automatically and publish 
+an `Event` so that the end user is notified regarding this switch to using the default generated
+certificates since the associated secret is not found. The `.status` is also updated to 
+`RouteCertificateDefaultType`.
+
+
+When transitioning from using a managed certificate to manually managed certificate, the user
+is expected to only delete the annotation. If the previously used `.spec.tls` is not deleted
+by the user the Route will continue to use the same certificate data and `.status` is updated 
+to `RouteCertificateCustomType`.
+
+The new secret-injector controller will not reconcile Route CRs that are owned by the 
+ingress-to-route controller.
+
+### Risks and Mitigations
+
+The new controller that is going to be introduced for this enhancement will have the capability 
+to read secret objects and write the secret data into the route CR. There is precedence 
+already for such intrusive permissions required by the ingress-to-route controller. 
+
+### Drawbacks
+
+This introduces an inconsistency between Ingress and Route CRD with respect to how 
+a secret reference can be provided with the certificate data. The ingress has a field
+`.spec.tls.secretName` where as the Route will have an annotation.
+
+## Design Details
+
+### Open Questions [optional]
+
+1. Does this warrant making changes to the ingress-to-route controller in how 
+   `.spec.tls.secretName` is processed? Currently, the controller reads the secret
+   data and copies it over to the Route CR that is created. Since this enhancement
+   introduces a difference between managed certificate provided via secrets and  
+   manually provided secrets, this distinction isn't possible with Routes created
+   via Ingresses.
+
+   - Proposed change
+     The ingress-to-route controller will also process the same annotation and if 
+     present copies it over to the route that is created. This annotation will take
+     precedence over `.spec.tls.secretName` and this offers enough distinction between
+     third-party/manual certificate management.
+
+**Answer**: The behaviour of the ingress-to-route controller will be retained as is and 
+no changes will be done.
```

Comment:
Would be nice to document the reason why.

Reply:
Addressed.